lib/IDE/CompletionInstance.cpp

```diff
@@ -176,85 +173,124 @@ bool CompletionInstance::performCachedOperaitonIfPossible(
+    if (!DC)
+      return false;
+
+    // OK, we can perform fast completion for this. Update the orignal delayed
+    // decl state.
+
+    // Fast completion keeps the buffer in memory for multiple completions.
+    // To reduce the consumption, slice the source buffer so it only holds
+    // the portion that is needed for the second pass.
+    auto startOffset = newInfo.StartOffset;
+    if (newInfo.PrevOffset != ~0u)
+      startOffset = newInfo.PrevOffset;
+    auto startLoc = tmpSM.getLocForOffset(tmpBufferID, startOffset);
+    startLoc = Lexer::getLocForStartOfLine(tmpSM, startLoc);
+    startOffset = tmpSM.getLocOffsetInBuffer(startLoc, tmpBufferID);
+
+    auto endOffset = newInfo.EndOffset;
+    auto endLoc = tmpSM.getLocForOffset(tmpBufferID, endOffset);
+    endLoc = Lexer::getLocForEndOfToken(tmpSM, endLoc);
+    endOffset = tmpSM.getLocOffsetInBuffer(endLoc, tmpBufferID);
+
+    newInfo.StartOffset -= startOffset;
+    newInfo.EndOffset -= startOffset;
+    if (newInfo.PrevOffset != ~0u)
+      newInfo.PrevOffset -= startOffset;
+
+    auto sourceText =
+        completionBuffer->getBuffer().slice(startOffset, endOffset);
+    auto newOffset = Offset - startOffset;
+
+    newBufferID = SM.addMemBufferCopy(sourceText,
+                                      completionBuffer->getBufferIdentifier());
+    SM.openVirtualFile(SM.getLocForBufferStart(newBufferID),
+                       tmpSM.getDisplayNameForLoc(startLoc),
+                       tmpSM.getLineAndColumn(startLoc).first - 1);
+    SM.setCodeCompletionPoint(newBufferID, newOffset);
+
+    // Construct dummy scopes. We don't need to restore the original scope
+    // because they are probably not 'isResolvable()' anyway.
+    auto &SI = oldState.getScopeInfo();
+    assert(SI.getCurrentScope() == nullptr);
+    Scope Top(SI, ScopeKind::TopLevel);
+    Scope Body(SI, ScopeKind::FunctionBody);
+
+    oldInfo.ParentContext = DC;
+    oldInfo.StartOffset = newInfo.StartOffset;
+    oldInfo.EndOffset = newInfo.EndOffset;
+    oldInfo.PrevOffset = newInfo.PrevOffset;
+    oldState.restoreCodeCompletionDelayedDeclState(oldInfo);
+
+    auto *AFD = cast<AbstractFunctionDecl>(DC);
+    if (AFD->isBodySkipped())
+      AFD->setBodyDelayed(AFD->getBodySourceRange());
+
+    break;
+  }
+  case CodeCompletionDelayedDeclKind::Decl:
+  case CodeCompletionDelayedDeclKind::TopLevelCodeDecl: {
+    // Support decl/top-level code only if the completion happens in a single
+    // file 'main' script (e.g. playground).
+
+    auto *M = oldInfo.ParentContext->getParentModule();
+    if (M->getFiles().size() != 1 || oldSF->Kind != SourceFileKind::Main)
+      return false;
+
+    // Re-parse the whole file. We still re-use imported modules.
+
+    auto sourceText = completionBuffer->getBuffer();
+    if (newInfo.Kind == CodeCompletionDelayedDeclKind::TopLevelCodeDecl) {
+      // We don't need the source text after the top-level code.
```

Comment:
How does this work for a whole SourceFile?  Won't slicing the content cause the interface hash to change if you move the cursor past a decl?

Reply:
If the completion happens in non-function body position, we don't take interface hash into account. We just re-parse the whole file. So I believe it shouldn't be a problem.